zkvm/src/types.rs

```diff
@@ -329,6 +375,36 @@ impl Data {
             },
         }
     }
+
+    /// Encodes blinded Data values for txprogram bytecode.
+    pub fn encode(&self, program: &mut Vec<u8>) {
+        match self {
+            Data::Opaque(x) => {
+                program.append(&mut x.clone());
+                return;
+            }
+            Data::Witness(w) => w.encode(program),
+        };
+    }
+}
+
+impl DataWitness {
+    fn encode(&self, program: &mut Vec<u8>) {
+        match self {
+            DataWitness::Program(instr) => Instruction::encode_program(instr.iter(), program),
+            DataWitness::Predicate(pw) => pw.encode(program),
+            DataWitness::Commitment(cw) => cw.encode(program),
+            DataWitness::Scalar(s) => program.extend_from_slice(&s.to_bytes()),
+            DataWitness::Input(b) => {
+                // Input = PreviousTxID || PreviousOutput
+                let (contract, _) = (&b.0, b.1);
+                // TBD: get prev_txid
```

Comment:
for the incorrect code it's better to slap `unimplemented!()` where the value is incorrect, so it's not sounding like "this is good-to-have optimization, can do it ~later~ never"

Reply:
Seems like it's conceptually cleaner to keep `Input` have the same structure as what we need to encode:

```
pub enum Input {
    Opaque(Vec<u8>),
    Witness(Box<(FrozenContract, TxID)>),
}
```

Or even a triplet:  contract, utxo, txid to avoid recomputing utxo by serializing a contract.